File: 01-arrays_&_hashing/00049-Group_Anagrams/00049-group_anagrams.java
// 49: Group Anagrams
// https://leetcode.com/problems/group-anagrams/

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    // SOLUTION
    public List<List<String>> groupAnagrams(String[] strs) {
        if (strs == null || strs.length == 0) return new ArrayList<>();

        Map<String, List<String>> tmap = new HashMap<>();

        for (String s : strs) {
            char[] ca = new char[26];
            for (char c : s.toCharArray()) ca[c - 'a']++;
            String keyStr = String.valueOf(ca);
            if (!tmap.containsKey(keyStr)) tmap.put(keyStr, new ArrayList<>());
            tmap.get(keyStr).add(s);
        }

        return new ArrayList<>(tmap.values());
    }

    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        String[] strs = {"eat","tea","tan","ate","nat","bat"};

        // OUTPUT
        var result = o.groupAnagrams(strs);
        System.out.print(result);
    }
}

File: 01-arrays_&_hashing/00049-Group_Anagrams/00049-group_anagrams.cpp
// 49: Group Anagrams
// https://leetcode.com/problems/group-anagrams/

#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution {
public:
    // SOLUTION
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> result;
        unordered_map<string, vector<string>> tmap;
        
        for(int  i=0; i<strs.size(); i++) {
            string str = strs[i];
            sort(strs[i].begin(), strs[i].end());
            tmap[strs[i]].push_back(str);
        }
        
        for(auto i=tmap.begin(); i!=tmap.end(); i++)
            result.push_back(i->second);
        
        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<string> strs = {"eat","tea","tan","ate","nat","bat"};

    // OUTPUT
    auto result = o.groupAnagrams(strs);
    cout<<"[";
    for (auto g : result) {
        cout<< "[";
        for (auto s : g) {
            cout << "\"" << s << "\",";
        }
        cout << "\b],";
    }
    cout << "\b]" << endl;

    return 0;
}
